core/src/main/java/com/yahoo/oak/EntryOrderedSet.java

```diff
@@ -299,7 +299,9 @@ boolean isEntrySetValidAfterRebalance() {
 
     void releaseAllDeletedKeys() {
         KeyBuffer key = new KeyBuffer(config.keysMemoryManager.getEmptySlice());
-        for (int i = 0; i < numOfEntries.get() ; i++) {
+        for (int i = 0; i < nextFreeIndex.get() - 1 && i < numOfEntries.get() ; i++) {
```

Comment:
This looks to me very weird condition and comment.
1. `releaseAllDeletedKeys()` should be called only by one thread that succeeds to change the chunk's state from FROZEN to RELEASED (CAS success). I do not see any reason for other threads to invoke `releaseAllDeletedKeys()`. Can you please check that this is how it happens? And definitely add a header comment about `releaseAllDeletedKeys()` usage intention.
2. This release happens as a very last step in the Chunk rebalance. At the beginning of the rebalance step the thread proceeding with the rebalance waits for all updates to finish. This is why we have publish/unpublish staff. So if `nextFreeIndex` is concurrently updated this is a bug... Maybe to add an explanatory comment here and simplify the loop condition. I think checking only `nextFreeIndex` should be enough.